pallets/messages/src/types.rs

```diff
@@ -26,6 +26,8 @@ where
 	///  Message source account id (the original source).
 	pub msa_id: Option<MessageSourceId>,
 	///  Stores index of message in block to keep total order.
+	// Is this index really needed, messages don't end up ordered in the Messages strorage Item (form on_initialize)?
+	// Maybe necessary for reading purposes from the StorageDoubleMap I guess?
```

Comment:
While the final storage is ordered, it is nested by `SchemaId`, so to be able to tell the ordering difference between different schemas, we need to keep this value.

Also if we modify the temp block, it requires modifying that memory.

@aramikm to double check. We might be able to remove it now.

Reply:
@NachoPal @wilwade 
The decision to use the temporary storage comes exactly from having canonical order. If we don't use the temporary storage , to be able to keep the total index we will need **1** additional `DB_READ` and **1** additional `DB_WRITE` per `add_message`. This is going to be very expensive since we are expecting that this extrinsic is going to be used a lot. By using the temporary storage we are getting around that additional read and write but we are going to have additional S (number of unique schemas used) DB_WRITES. Since S is always expected to be smaller than total number of added messages in a block, from weights perspective it makes sense to do this process `on_initialize`.

If we don't need total order then we can get rid of this temporary storage and directly add to `Message` storage

Reply:
Based on my comparisons between these two different implementations we are getting the following equation

M = number of messages in a block
S = Number of unique schemas used in block
DB_READ = 25_000_000
DB_WRITE = 100_000_000

`
 121_000_000 M > 185_619_000 S`

so if we assume S = 5 then M > 7
which means after adding the 7's message in block the weights of regular implementation will become bigger than our current implementation
and for S = 20 we get M > 30
and ...